code/modules/wiremod/port.dm

```diff
@@ -52,185 +53,165 @@
- * and applying/removing signals on inputs
+ * Sets the port's value to v.
+ * Casts to the port's datatype (e.g. number -> string), and assumes this can be done.
  */
-/datum/port/proc/convert_value(prev_value, value_to_convert)
-	if(prev_value == value_to_convert)
-		return prev_value
-	. = value_to_convert
-
-	if(isnull(value_to_convert))
-		return null
+/datum/port/proc/put(v)
+	if(value == v)
+		return
+	if(isatom(value))
+		UnregisterSignal(value, COMSIG_PARENT_QDELETING)
+	value = v
+	if(isnull(value))
+		return
+	value = cast(value, datatype)
+	if(isatom(value))
+		var/atom/atom_to_check = value
+		if(QDELETED(atom_to_check))
+			return null
+		RegisterSignal(value, COMSIG_PARENT_QDELETING, .proc/null_output)
 
-	switch(datatype)
+/**
+ * Implicit conversion of a value to a type.
+ * Assumes that the types are compatible.
+ */
+/proc/cast(value, type)
+	switch(type)
 		if(PORT_TYPE_STRING)
 			// So that they can't easily get the name like this.
-			if(isatom(value_to_convert))
+			if(isatom(value))
 				return PORT_TYPE_ATOM
-			else
-				return copytext("[value_to_convert]", 1, PORT_MAX_STRING_LENGTH)
-
-	if(isatom(value_to_convert))
-		var/atom/atom_to_check = value_to_convert
-		if(QDELETED(atom_to_check))
-			return null
+			return copytext("[value]", 1, PORT_MAX_STRING_LENGTH)
+	return value
 
 /**
  * Sets the datatype of the port.
  *
  * Arguments:
- * * type_to_set - The type this port is set to.
+ * * new_type - The type this port is to be set to.
  */
-/datum/port/proc/set_datatype(type_to_set)
-	datatype = type_to_set
+/datum/port/proc/set_datatype(new_type)
+	var/old_type = datatype
+	datatype = new_type
+	put(compatible_datatypes(old_type, new_type) ? value : null)
 	color = datatype_to_color()
-	disconnect()
 	if(connected_component?.parent)
 		SStgui.update_uis(connected_component.parent)
 
-/**
- * Disconnects a port from all other ports
- *
- * Called by [/obj/item/circuit_component] whenever it is disconnected from
- * an integrated circuit
- */
-/datum/port/proc/disconnect()
-	SHOULD_CALL_PARENT(TRUE)
-	SEND_SIGNAL(src, COMSIG_PORT_DISCONNECT)
+/datum/port/output/set_datatype(new_type)
+	SEND_SIGNAL(src, COMSIG_WIREMOD, new_type=new_type)
+	..()
 
+/datum/port/input/set_datatype(new_type)
+	for(var/datum/port/output/output as anything in connected_ports)
+		if(!compatible_datatypes(output.datatype,new_type))
+			disconnect(output)
+	..()
 
 /**
- * # Output Port
- *
- * An output port that many input ports can connect to
- *
- * Sends a signal whenever the output value is changed
+ * Disconnects a port from all other ports.
  */
-/datum/port/output
-	/// The output value of the port
-	var/output_value
+/datum/port/proc/disconnect_all()
 
-/datum/port/output/disconnect()
-	set_output(null)
-	return ..()
+/datum/port/input/disconnect_all()
+	for(var/datum/port/port as anything in connected_ports)
+		disconnect(port)
+
+/datum/port/output/disconnect_all()
+	SEND_SIGNAL(src, COMSIG_WIREMOD, disconnect=TRUE)
```

Comment:
BAD. This is NOT how signals work.

Signals should be fired from *one* source, and should be *events*, not proc-calls.